tracer/src/Datadog.Trace/Configuration/TracerSettingsSnapshot.cs

```diff
@@ -0,0 +1,328 @@
+﻿// <copyright file="TracerSettingsSnapshot.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Datadog.Trace.Configuration.Telemetry;
+
+namespace Datadog.Trace.Configuration;
+
+/// <summary>
+/// Holds a snapshot of the mutable <see cref="TracerSettings"/> values after
+/// initial creation. Used for tracking which properties have been changed in code
+/// subsequently. Only properties which can be modified externally (both publicly, and
+/// outside the mutable settings objects should be recorded. Mutation of settings
+/// _inside_ <see cref="TracerSettings"/> et al should be explicitly tracked when
+/// they are mutated.
+/// </summary>
+internal class TracerSettingsSnapshot
+{
+    private static readonly HashSet<string> EmptyHashSet = new();
+
+    public TracerSettingsSnapshot(TracerSettings settings)
+    {
+        Environment = settings.EnvironmentInternal;
```

Comment:
Rather than having all those properties, wouldn't it be better to create and store an instance of ImmutableTracerSettings as snapshot?

Reply:
Yeah, it's an option, but `ImmutableTracerSettings` has 55 properties, whereas this only has ~20 which is less alocatey by my book 😉  It also means we don't allocate empty dictionaries etc if we don't need them

Reply:
Went with a half-way approach and source generated the snapshot object instead 😄